File: Game_Of_Life_Python/Organisms/Animals/turtle.py
from Organisms.organism import OrganismInitialData
from Organisms.animal import Animal, CollisionTypes
import random


class Turtle(Animal):

    static_counter = 0

    def __init__(self, name, row, column, given_strength=-1, given_initiative=-1):
        strength = OrganismInitialData[name]["strength"]
        initiative = OrganismInitialData[name]["initiative"]
        character = OrganismInitialData[name]["character"]
        if given_strength != -1:
            strength = given_strength
        if given_initiative != -1:
            initiative = given_initiative
        super().__init__(strength, initiative, name, character, row, column, "turtle.png")
        Turtle.static_counter += 1
        self.type = Turtle

    def action(self, grid_board):
        rand = random.randint(1, 4)
        if (rand == 4):  # Means turtle move
            self.default_action_animal(grid_board)
        else:
            self.print_to_journal(
                f"{self.name} stays at ({self.row}, {self.column})\n")

    def collision(self, grid_board, organisms, current_index):
        return self.default_collision_animal(grid_board, organisms, current_index)

    def get_static_counter(self):
        return Turtle.static_counter

File: GameOfLife_Cpp/source/src/turtle.cpp
#include "turtle.hpp"
#include <iostream>

int Turtle::TURTLE_STATIC_COUNTER = 0;

Turtle::Turtle(int row, int column)
    : Animal(2, 1, 0, "turtle", 't', row, column, "turtle.png", OrganismType::TURTLE)
{
    TURTLE_STATIC_COUNTER++;  
}

Turtle::Turtle(int row, int column, int strength, int initiative, int age)
    : Animal(strength, initiative, age, "turtle", 't', row, column, "turtle.png", OrganismType::TURTLE)
{
    TURTLE_STATIC_COUNTER++;  
    this->get_age() = age;
}

Turtle::~Turtle()
{
    TURTLE_STATIC_COUNTER--;
}

ActionResult Turtle::action(std::vector<std::vector<char>>& grid_board)
{
    previous_row = row;
    previous_column = column;
    int random = rand() % 4 + 1;
    if (random == 4)
        return Animal::default_action_animal(grid_board);
    std::cout << this->get_name() << " stays at the same position: (" << row << ", " << column << ")\n";
    return ActionResult(ActionType::STAY);
}

CollisionResult Turtle::collision(std::vector<std::vector<char>>& grid_board, std::vector<Organism*>& organisms, int current_index)
{
    return this->default_collision_animal(grid_board, organisms, current_index);
}

int Turtle::get_static_counter()
{
    return TURTLE_STATIC_COUNTER;
}